contracts/pangolin-periphery/PangolinBridgeMigrationRouter.sol

```diff
@@ -0,0 +1,358 @@
+        bridgeMigrator[token] = migrator;
+    }
+
+    /**
+     * @notice Internal function to manage approval, allows an ERC20 to be spent to the maximum
+     * @param tokenAddress The ERC20 token address that will be allowed to be used
+     * @param spenderAddress Who's going to spend the ERC20 token
+     */
+    function _allowToken(address tokenAddress, address spenderAddress) internal {
+        IPangolinERC20(tokenAddress).approve(spenderAddress, uint(-1));
+    }
+
+    /**
+     * @notice Internal function add liquidity on a low level, without the use of a router
+     * @dev This function will try to maximize one of the tokens amount and match the other
+     * one, can cause dust so consider charge backs
+     * @param pairToken The pair that will receive the liquidity
+     * @param token0 The first token of the pair
+     * @param token1 The second token of the pair
+     * @param amountIn0 The amount of first token that can be used to deposit liquidity
+     * @param amountIn1 The amount of second token that can be used to deposit liquidity
+     * @param to The address who will receive the liquidity
+     * @return amount0 Charge back from token0
+     * @return amount1 Charge back from token1
+     * @return liquidityAmount Total liquidity token amount acquired
+     */
+    function _addLiquidity(
+        address pairToken,
+        address token0,
+        address token1,
+        uint amountIn0,
+        uint amountIn1,
+        address to
+    ) private returns (uint amount0, uint amount1, uint liquidityAmount) {
+        (uint112 reserve0, uint112 reserve1,) = IPangolinPair(pairToken).getReserves();
+        uint quote0 = amountIn0;
+        uint quote1 = PangolinLibrary.quote(amountIn0, reserve0, reserve1);
+        if (quote1 > amountIn1) {
+            quote1 = amountIn1;
+            quote0 = PangolinLibrary.quote(amountIn1, reserve1, reserve0);
+        }
+        TransferHelper.safeTransfer(token0, pairToken, quote0);
+        TransferHelper.safeTransfer(token1, pairToken, quote1);
+        amount0 = amountIn0 - quote0;
+        amount1 = amountIn1 - quote1;
+        liquidityAmount = IPangolinPair(pairToken).mint(to);
+    }
+
+    /**
+     * @notice Internal function to remove liquidity on a low level, without the use of a router
+     * @dev This function requires the user to approve this contract to transfer the liquidity pair on it's behalf
+     * @param liquidityPair The pair that will have the liquidity removed
+     * @param amount The amount of liquidity token that you want to rescue
+     * @return amountTokenA The amount of token0 from burning liquidityPair token
+     * @return amountTokenB The amount of token1 from burning liquidityPair token
+     */
+    function _rescueLiquidity(
+        address liquidityPair,
+        uint amount
+    ) internal returns (uint amountTokenA, uint amountTokenB) {
+        TransferHelper.safeTransferFrom(liquidityPair, msg.sender, liquidityPair, amount);
```

Comment:
double check whether this could be vulnerable to a re-entrancy attack

Reply:
Same as above